### PR TITLE
support for internal validation against theoretical values

### PR DESCRIPTION
**Pull request by abilandz**

### CODE REVIEW

File: PWGCF/MultiparticleCorrelations/Core/MuPa-DataMembers.h [hunk 1]
```diff
@@ -12,11 +12,13 @@
...
-// 0. Starting with C++11, it's possible to initialize data members at
-// declaration, so I do it here
-// 1. Use //!<! for introducing a Doxygen comment interpreted as transient in
-// both ROOT 5 and ROOT 6.
+// General remarks:
+// 0. Starting with C++11, it's possible to initialize data members at declaration, so I do it here
+// 1. Use //!<! for introducing a Doxygen comment interpreted as transient in both ROOT 5 and ROOT 6.
+// 2. If I hit at the compilation error: Framework/StructToTuple.h:286:6: error: only 99 names provided for structured binding
+//    2a) As far as I can tell, that means that sum of individual data members + struct fields + individual configurables > 99
+//    2b) Therefore, wrap up all data members in some struct fields + use in instead of individual configurables arrays whenever possible.
```
Comment (victor-gonzalez) [inline PWGCF/MultiparticleCorrelations/Core/MuPa-DataMembers.h:20]: Consider the possibility of using `ConfigurableGroup`

**Review by victor-gonzalez — APPROVED** [review]
I fixed the warning but have a look at my other comments

File: PWGCF/MultiparticleCorrelations/Core/MuPa-DataMembers.h [hunk 2]
```diff
@@ -113,13 +102,16 @@ struct EventCuts {
...
+  TProfile* fParticleHistogramsPro = NULL;                                    //!<! keeps flags relevant for the control particle histograms
+  TH1D* fParticleHistograms[eParticleHistograms_N][2][2] = {{{NULL}}};        //! [ type - see enum eParticleHistograms ][reco,sim][before, after particle cuts]
+  Bool_t fBookParticleHistograms[eParticleHistograms_N] = {kTRUE};            // book or not this histogram, see configurable cfBookParticleHistograms
+  Double_t fParticleHistogramsBins[eParticleHistograms_N][3] = {{0.}};        // [nBins,min,max]
+  Double_t fParticleCuts[eParticleHistograms_N][2] = {{0.}};                  // [min,max]
+  TH2D* fParticleHistograms2D[eParticleHistograms2D_N][2][2] = {{{NULL}}};    //! [ type - see enum eParticleHistograms2D ][reco,sim][before, after particle cuts]
+  Bool_t fBookParticleHistograms2D[eParticleHistograms2D_N] = {kTRUE};        // book or not this 2D histogram, see configurable cfBookParticleHistograms2D
+  Double_t fParticleHistogramsBins2D[eParticleHistograms2D_N][2][3] = {{0.}}; // [type - see enum][x,y][nBins,min,max]
```
[the remark this answers is not in the document]
Comment (abilandz) [reply PWGCF/MultiparticleCorrelations/Core/MuPa-DataMembers.h]: Thanks! I checked in general if dimensionality is correct at initialization, but I overlooked this one...

File: PWGCF/MultiparticleCorrelations/Core/MuPa-DataMembers.h [hunk 3]
```diff
@@ -135,7 +127,7 @@ struct Qvector {
   Bool_t fCalculateQvectors = kTRUE;                                                                                                 // to calculate or not to calculate Q-vectors, that's a Boolean...
   TComplex fQ[gMaxHarmonic * gMaxCorrelator + 1][gMaxCorrelator + 1] = {{TComplex(0., 0.)}};                                         //! generic Q-vector
   TComplex fQvector[gMaxHarmonic * gMaxCorrelator + 1][gMaxCorrelator + 1] = {{TComplex(0., 0.)}};                                   //! "integrated" Q-vector
-  TComplex fqvector[eqvectorKine_N][gMaxNoBinsKine][gMaxHarmonic * gMaxCorrelator + 1][gMaxCorrelator + 1] = {{{{TComplex(0., 0.)}}}}; //! "differenttial" q-vector [kine var.][binNo][fMaxHarmonic*fMaxCorrelator+1][fMaxCorrelator+1] = [6*12+1][12+1]
+  TComplex fqvector[eqvectorKine_N][gMaxNoBinsKine][gMaxHarmonic * gMaxCorrelator + 1][gMaxCorrelator + 1] = {{{TComplex(0., 0.)}}}; //! "differenttial" q-vector [kine var.][binNo][fMaxHarmonic*fMaxCorrelator+1][fMaxCorrelator+1] = [6*12+1][12+1]
```
[the remark this answers is not in the document]
Comment (abilandz) [reply PWGCF/MultiparticleCorrelations/Core/MuPa-DataMembers.h]: Thanks!

File: PWGCF/MultiparticleCorrelations/Tasks/multiparticle-correlations-ab.cxx [hunk 4]
```diff
@@ -209,7 +212,7 @@ struct MultiparticleCorrelationsAB // this name is used in lower-case format to
   // -------------------------------------------
 
   // F) Process only converted simulated Run 2 data:
-  void processSim_Run2(aod::Collision const&) // TBI 20240224 not ready yet, this is just dummy to version to get later "doprocess..." variable
+  void processSim_Run2(aod::Collision const& collision) // TBI 20240224 not ready yet, this is just dummy to version to get later "doprocess..." variable
```
[the remark this answers is not in the document]
Comment (abilandz) [reply PWGCF/MultiparticleCorrelations/Tasks/multiparticle-correlations-ab.cxx]: Thanks, noted!

File: PWGCF/MultiparticleCorrelations/Core/MuPa-MemberFunctions.h [hunk 5]
```diff
@@ -42,37 +42,41 @@ void BookBaseList()
                                     Form("fTaskName = %s", tc.fTaskName.Data()));
 
   fBasePro->GetXaxis()->SetBinLabel(eRunNumber,
-                                    Form("fRunNumber = %s", tc.fRunNumber.Data()));
+                                    Form("fRunNumber = %s", "__RUN_NUMBER__")); // I have to do it this way via placeholder, because run number is available only when i start to process date
+                                                                                // Then, I replace placeholder with run number in DetermineAndPropagateRunNumber(T const& collision)
 
   fBasePro->GetXaxis()->SetBinLabel(eDryRun, "fDryRun");
-  fBasePro->Fill(+eDryRun - 0.5, (Int_t)tc.fDryRun);
+  fBasePro->Fill(static_cast<float>(eDryRun) - 0.5, static_cast<int>(tc.fDryRun));
```
[the remark this answers is not in the document]
Comment (abilandz) [reply PWGCF/MultiparticleCorrelations/Core/MuPa-MemberFunctions.h:49]: This reflects my personal fear of filling the bin with the value that falls exactly on the edge of that bin. That explains the offset -0.5, which shall ensure that I fill in the bin center. Then, I do not have to worry if the lower or the upper boundary of the bin is included in the range of that bin. But I will re-think the implementation here.

File: PWGCF/MultiparticleCorrelations/Core/MuPa-MemberFunctions.h [hunk 6]
```diff
@@ -1213,16 +1350,346 @@ void BookNestedLoopsHistograms()
...
+    LOGF(fatal, "in function \033[1;31m%s at line %d : !planes \n \033[0m", __FUNCTION__, __LINE__);
+  }
+  if (amplitudes->GetSize() != planes->GetSize()) {
+    LOGF(fatal, "in function \033[1;31m%s at line %d : amplitudes->GetSize() != planes->GetSize() \n \033[0m", __FUNCTION__, __LINE__);
+  }
+
+  // b) Main calculus:
+  TComplex value = TComplex(1., 0., kTRUE); // yes, polar representation
+  for (Int_t h = 0; h < harmonics->GetSize(); h++) {
+    // cout<<"h = "<<h<<" ... "<<TMath::Abs(harmonics->GetAt(h))<<endl;
```
[the remark this answers is not in the document]
Comment (abilandz) [reply PWGCF/MultiparticleCorrelations/Core/MuPa-MemberFunctions.h:1480]: OK, I overlooked this one!

File: PWGCF/MultiparticleCorrelations/Core/MuPa-MemberFunctions.h [hunk 7]
```diff
@@ -1864,47 +2388,64 @@ Bool_t ValidTrack(T const& track)
...
+  // c) Additional validity checks for all tracks (in Run 3, 2 and 1), use only during debugging:
+  if (tc.fInsanityCheckForEachParticle) {
+
+    // *) isnan() check (remember that 'nan' is 0./0., inf-inf, etc. However 'inf' itself is NOT a 'nan', therefore isnan(1./0.) is false, isnan(0./0.) is true, etc.):
     if (isnan(track.phi()) || isnan(track.pt()) || isnan(track.eta())) {
-      LOGF(error, "\033[1;33m%s nan hit, skipping this track \033[0m", __PRETTY_FUNCTION__);
-      LOGF(error, "dPhi = %f\ndPt = %f\ndEta = %f\ntrackType() = %d", track.phi(), track.pt(), track.eta(), static_cast<int>(track.trackType())); // TBI 20240226 check if I need this cast here
+      if (tc.fVerboseForEachParticle) {
+        LOGF(info, "\033[1;31m%s isnan(track.phi()) || isnan(track.pt()) || isnan(track.eta())\033[0m", __FUNCTION__);
+        LOGF(error, "track.phi() = %f\ntrack.pt() = %f\ntrack.eta() = %f", track.phi(), track.pt(), track.eta());
```
[the remark this answers is not in the document]
Comment (abilandz) [reply PWGCF/MultiparticleCorrelations/Core/MuPa-MemberFunctions.h:2435]: I have to invest some effort to make all info, error and fatal log messages internally self-consistent. Then I will address this one as well, it won't stay as it is now.

File: PWGCF/MultiparticleCorrelations/Core/MuPa-DataMembers.h (hunk 1 above)
Comment (abilandz) [reply PWGCF/MultiparticleCorrelations/Core/MuPa-DataMembers.h:20]: Thanks, I was not aware of that possibility.

**Pull request merged**